File: SQL Server 2012/Exercises/Chapter 23/C#/ApClrObjectsEx/ApClrObjectsEx/WarehouseLocation.cs
//------------------------------------------------------------------------------
// <copyright file="CSSqlUserDefinedType.cs" company="Microsoft">
//     Copyright (c) Microsoft Corporation.  All rights reserved.
// </copyright>
//------------------------------------------------------------------------------
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;


[Serializable]
[Microsoft.SqlServer.Server.SqlUserDefinedType(
    Format.UserDefined,
    MaxByteSize = 5)]
public struct WarehouseLocation : INullable, IBinarySerialize
{
    private bool isNull;
    private string location;

    public override string ToString()
    {
        return location;
    }
    
    public bool IsNull
    {
        get
        {
            return isNull;
        }
    }
    
    public static WarehouseLocation Null
    {
        get
        {
            WarehouseLocation warehouseLocation = new WarehouseLocation();
            warehouseLocation.isNull = true;
            return warehouseLocation;
        }
    }

    public SqlString Location
    {
        get
        {
            return location;
        }
    }

    public SqlString Rack
    {
        get
        {
            return location.Substring(0, 1);
        }

    }

    public SqlString Shelf
    {
        get
        {
            return location.Substring(1, 1);
        }
    }

    public SqlString Bin
    {
        get
        {
            return location.Substring(2);
        }
    }

    public static WarehouseLocation Parse(SqlString s)
    {
        if (s.IsNull)
            return Null;

        Validate(s);

        WarehouseLocation warehouseLocation = new WarehouseLocation();
        warehouseLocation.location = s.ToString();
        return warehouseLocation;
    }

    // user-defined data validation method

    private static void Validate(SqlString str)
    {
        string s = str.Value;
        string message = "Invalid email: " + s + ". ";

        // check that the rack is a letter
        Char rack = s[0];
        if (!Char.IsLetter(rack))
            throw new ApplicationException(message +
                "The rack must be alphabetic.");

        // check that the shelf is an integer between 1 and 5
        try
        {
            int shelf = Convert.ToInt32(s.Substring(1, 1));
            if (shelf < 1 || shelf > 5)
                throw new ApplicationException(message +
                    "The shelf must be between 1 and 5.");
        }
        catch (FormatException)
        {
            throw new ApplicationException(message +
                "The shelf must be an integer.");
        }

        // check that the bin is an integer between 1 and 50
        try
        {
            int bin = Convert.ToInt32(s.Substring(2));
            if (bin < 1 || bin > 50)
                throw new ApplicationException(message +
                    "The bin must be between 1 and 50.");
        }
        catch (FormatException)
        {
            throw new ApplicationException(message +
                "The bin must be an integer.");
        }

    }
    
    public void Read(System.IO.BinaryReader r)
    {
        location = r.ReadString();
    }

    public void Write(System.IO.BinaryWriter w)
    {
        w.Write(location);
    }
}

File: SQL Server 2012/Exercises/Chapter 23/Visual Basic/ApClrObjectsEx/WarehouseLocation.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server

<Serializable()> _
<Microsoft.SqlServer.Server.SqlUserDefinedType(
    Format.UserDefined,
    MaxByteSize:=5)> _
Public Structure WarehouseLocation
    Implements INullable, IBinarySerialize

    Private m_isNull As Boolean
    Private m_location As String

    Public Overrides Function ToString() As String
        Return m_location
    End Function

    Public ReadOnly Property IsNull() As Boolean Implements INullable.IsNull
        Get
            Return m_isNull
        End Get
    End Property

    Public Shared ReadOnly Property Null() As WarehouseLocation
        Get
            Dim location As New WarehouseLocation
            location.m_isNull = True
            Return location
        End Get
    End Property

    Public ReadOnly Property Location() As SqlString
        Get
            Return m_location
        End Get
    End Property

    Public ReadOnly Property Rack() As SqlString
        Get
            Return m_location.Substring(0, 1)
        End Get
    End Property

    Public ReadOnly Property Shelf() As SqlString
        Get
            Return m_location.Substring(1, 1)
        End Get
    End Property

    Public ReadOnly Property Bin() As SqlString
        Get
            Return m_location.Substring(2)
        End Get
    End Property

    Public Shared Function Parse(ByVal s As SqlString) As WarehouseLocation
        If s.IsNull Then
            Return Null
        End If

        Validate(s)

        Dim location As New WarehouseLocation
        location.m_location = s.ToString
        Return location
    End Function

    Private Shared Sub Validate(ByVal str As SqlString)
        Dim s As String = str.Value
        Dim message As String = "Invalid location: " & s & ". "

        Dim rack As Char = s(0)
        If Not Char.IsLetter(rack) Then
            Throw New ApplicationException(message &
                "The rack must be alphabetic.")
        End If

        Try
            Dim shelf As Integer = Convert.ToInt32(s.Substring(1, 1))
            If shelf < 1 OrElse shelf > 5 Then
                Throw New ApplicationException(message &
                    "The shelf must be between 1 and 5.")
            End If
        Catch ex As FormatException
            Throw New ApplicationException(message &
                "The shelf must be an integer.")
        End Try

        Try
            Dim bin As Integer = Convert.ToInt32(s.Substring(2))
            If bin < 1 OrElse bin > 50 Then
                Throw New ApplicationException(message &
                    "The bin must be between 1 and 50.")
            End If
        Catch ex As FormatException
            Throw New ApplicationException(message &
                "The bin must be an integer.")
        End Try

    End Sub

    Public Sub Read(ByVal r As System.IO.BinaryReader) _
            Implements Microsoft.SqlServer.Server.IBinarySerialize.Read
        m_location = r.ReadString
    End Sub

    Public Sub Write(ByVal w As System.IO.BinaryWriter) _
            Implements Microsoft.SqlServer.Server.IBinarySerialize.Write
        w.Write(m_location)
    End Sub

End Structure

